README.md
# node-occ

OpenCascade nodejs extension for solid modeling.

This nodejs extension provides *solid construction* to nodejs. 
It provides a simple yet powerful javascript api to construct 3D geometry models.

This project comes with a set of V8 wrappers around OpenCascade API and a sample web application.


### quick example


```javascript
var occ = require('occ')

// construct a box
var box = occ.makeBox([0,0,0],[100,100,50])

// construct a cylinder
var cyl = occ.makeCylinder([50,50,-10],[50,50,60],40)

// cut the box with cylinder
box = occ.cut(box,cyl)

// save result to a STEP file
occ.writeSTEP("somefile.step",box)

```


### list of features

- creation of basic shapes ( box, cylinder , cone , torus )
- boolean operation ( fuse , common , cut )
- features  ( draftAngle)
- solid properties ( faces, edges, vertices, area , volume )
- import export ( STEP BREP )



### sample web application

''/sample'' : sample nodejs/express REST API server to build solid , based on threejs



## installing node-occ from npm
 
### on Windows
  - you need Microsoft Visual Studio 2012
  - please install OpenCascade oce-0.11  from https://github.com/tpaviot/oce/downloads to you c:\OCE-0.11 folder
  
### on Linux




## building node-occ :  prerequisites

### on ubuntu

installing node js
```bash

# installing nodejs and gyp utility to build extensions
sudo apt-get install nodejs npm
sudo npm install node-gyp -g
sudo npm install mocha -g

#ddd installing prebuild opencascade
#ddd sudo apt-get install oce-draw

------------------------------------
# building opencascade from oce
#installing cmake
sudo apt-get install cmake cmake-curses-gui
# fetching source code
git clone https://github.com/tpaviot/oce.git
# switching to official release 0.11
cd oce
git checkout 0.11
cd ..
# creating a build area for oce
make build_oce
cd build_oce
cmake ../oce
ccmake .
#    turn on precompiled header compilation
#    remove X11

# building oce
make -j 4



git clone https://github.com/erossignon/node-occ.git
cd node-occ
npm install
node-gyp configure
node-gyp build
mocha

cd sample
npm install
node app.js
```


### on windows
  - please install oce-0.11  from https://github.com/tpaviot/oce/downloads to you c:\OCE-0.11 folder


### on Mac


## dependencies:

 - threejs    : https://github.com/mrdoob/three.js
 - codemirror : https://github.com/marijnh/CodeMirror

## acknowledgement:
    
 - OpenCascade : http://www.opencascade.org
 - occmodel    : https://github.com/tenko/occmodel
 - ShapeSmith  : https://github.com/bjnortier/shapesmith



## MIT License

Copyright © 2012 E. Rossignon

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.md

0.0.1  (2013-02-10)
===================

  * initial release
  
  

test1.js
var util   = require('util');
var assert = require('assert');

var occ = require('./');


function testOuterShell()
{
    var solid1 = occ.makeBox([10,20,30],[20,40,60]);
    var solid2 = occ.makeBox([15,25,35],[-20,-40,-60]);
    var solid = occ.fuse(solid1,solid2);


    solid1 = occ.makeBox([0,0,0],[20,20,20]);
    solid2 = occ.makeBox([10,10,10],[15,15,15]);
    solid = occ.cut(solid1,solid2);

    solid.getOuterShell();

    var shells = solid.getShells();
    for (var i in shells) {
        var shell = shells[i];
        console.log(shell.shapeType,shell.numFaces);
    }
}
testOuterShell();

console.log("-------------------");

function testCyl()
{
   var cyl =occ.makeCylinder([-100,20,30],[100,20,30],40);

}
testCyl();

var shapeFactory = require('./lib/shapeFactory.js');


var CSGbuilder =require('./lib/CSGbuilder');

function testCSG1()
{
    var solid = shapeFactory.makePan(occ);

}
testCSG1();

function testCSG()
{
    var simple_csg = {
        "box":    { type: "makeBox" , parameters:  [ [-10,-10,-10] , [10,10,10]] },
        "sphere": { type: "makeSphere" , parameters: [ [0,0,0] , 3 ] },
        "solid":  { type: "common", parameters: ["box","sphere"]  }
    };
    CSGbuilder.build(simple_csg);

}
testCSG();


var bbox = new occ.BoundingBox();
console.log(bbox.nearPt);
console.log(bbox.nearPt.x);

//var bottle = shapeFactory.makeBottle();

function test3()
{
    solid = occ.makeBox([10,20,30],[20,40,60]);
    solid = occ.makeFillet(solid,solid.getEdges(),2);

}
test3();

function test2(){
    var trsf ;
    trsf=new occ.Transformation();
    trsf.makePlaneMirror([10,20,30],occ.ZDir);

    console.log(" Scale Factor=",trsf.scaleFactor);

}
test2();

function test1() {
var solid;
var shapeIt;

solid = occ.makeBox([10,20,30],[20,40,60]);

shapeIt = new occ.ShapeIterator(solid,"FACE");
    shapeIt.more.should.be.true;
    assert(shapeIt.current === undefined);

}
test1();



var wire = new occ.Wire();
console.log(wire.numEdges);
console.log(wire.numVertices);

var trsf;
trsf = new occ.Transformation();
trsf.makePlaneMirror([10,20,30],occ.ZDir);




var v = new occ.Vertex();
var e = new occ.Edge().createLine([10,20,30],[10,14,15]);
console.log(e.length)

edge = new occ.Edge();
edge.createCircle([10,10,10],[0,0,1],20);


console.log("hello");

var box1 = new occ.Solid();

assert(box1.isNull);
assert(!box1.isValid);

box1 = occ.makeBox([10,20,30],[30,40,50]);

//xx for (var i in box1.mesh) {
//xx     console.log( " i=",i,"  => ", box1.mesh[i]);
//xx }
console.log("---------------!!!");

console.log( " i=",i,"  => ", box1.mesh.vertices);

assert(!box1.isNull);
assert(box1.isValid);

var box2 = occ.makeBox([10,20,30],[100,100,100]);
var box3 = occ.makeBox([10,20,30],100,100,100);

//xx console.log("box1.x",box1.location.x);
//xx console.log("box1.x",box1.location.y);
//xx console.log("box1.x",box1.location.z);
//xx console.log("cuboid",JSON.stringify(box1));


var solid = occ.makeBox([10,20,30],[20,30,40]);


for ( var i in solid) {
	console.log(" i = ", i, " => " , solid[i]);
}

var expected = ["isNull","area","volume","numFaces","numSolids","isValid","rotate","fuse"];

for ( var j in occ.Solid.prototype) {
	console.log(" j = ", j, " => " ,  occ.Solid.prototype[j]);
}	


console.log("------------------");



console.log("JSON =",solid.mesh.toJSON());


occ.writeSTEP("toto.STEP",solid);

console.log("Done");
